src/api/hapi/symbol_manager.cpp

```diff
@@ -0,0 +1,102 @@
+/*******************************************************
+ * Copyright (c) 2015, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include "symbol_manager.hpp"
+
+AFSymbolManager& AFSymbolManager::getInstance()
+{
+    static AFSymbolManager symbolManager;
+    return symbolManager;
+}
+
+/*flag parameter is not used on windows platform */
+LibHandle openDynLibrary(const char* dlName, int flag=RTLD_LAZY)
+{
+#if defined(OS_WIN)
+    HMODULE retVal = LoadLibrary(dlName);
+    if (retVal == NULL) {
+        retVal = LoadLibraryEx(dlName, NULL, LOAD_LIBRARY_SEARCH_DEFAULT_DIRS);
+    }
+    return retVal;
+#else
+    return dlopen(dlName, flag);
+#endif
+}
+
+void closeDynLibrary(LibHandle handle)
+{
+#if defined(OS_WIN)
+    FreeLibrary(handle);
+#else
+    dlclose(handle);
+#endif
+}
+
+AFSymbolManager::AFSymbolManager()
+    : isCPULoaded(false), isCUDALoaded(false), isOCLLoaded(false)
+{
+    cpuHandle = openDynLibrary(LIB_AF_CPU_NAME);
+    if (cpuHandle) {
+        isCPULoaded = true;
+        activeHandle = cpuHandle;
+    }
+    cudaHandle = openDynLibrary(LIB_AF_CUDA_NAME);
+    if (cudaHandle) {
+        isCUDALoaded = true;
+        activeHandle = cudaHandle;
+    }
+    oclHandle = openDynLibrary(LIB_AF_OCL_NAME);
+    if (oclHandle) {
+        isOCLLoaded = true;
+        activeHandle = oclHandle;
+    }
+}
```

Comment:
I didn't quite understand what is the choice here ? isn't that the reason af_set_backend function is there. In the even of such function is not called upon, activeHandle points to the backend in the order you just mentioned.